source/agora/consensus/protocol/Nominator.d

```diff
@@ -544,6 +553,21 @@ extern(D):
                     fail_reason, data);
                 return ValidationLevel.kInvalidValue;
             }
+
+            bool is_same = this.ledger.isSameMissingValidatorsAsLocal(data);
+            if (is_same)
+            {
+                if (auto fail_reason = this.ledger.validateSlashingData(data))
```

Comment:
`isSameMissingValidatorsAsLocal` and `validateSlashingData` (which forwards to `isInvalidPreimageRootReason`) seem to have similar checks. They both check the `missing_validators` array.

Actually thinking about this, maybe we can just remove `preimage_root` from the `ConsensusData`? The root can be easily calculated after consensus is reached, when the block is about to be created and signed. So we wouldn't have to check its validity here, we would just have to compare the two arrays. What do you think, is it a good idea?

In fact I think we only need to verify the root when downloading blocks from other nodes.

Reply:
> maybe we can just remove preimage_root from the ConsensusData?

I think we can do it because all the nodes have the same enrolled UTXOs (a.k.a. Validators) and we can calculate the root only with the indices. It's a good idea.

> In fact I think we only need to verify the root when downloading blocks from other nodes.

In fact, how to process the catch-up is a blocker for me now. There is some problem in checking the validity of the root in the middle of downloading blocks from other nodes. Suppose that the latest height is 1009 and all the current validators have enrolled at the height of 1008. When downloading the blocks between 0 and 1008, we can not check the validity of the root because there are no pre-images for un-enrolled validators.

Reply:
> there are no pre-images for un-enrolled validators.

That's why we should keep the last preimage and enrollment of every historic validator. It should be possible if we follow my new suggestion here: https://github.com/bpfkorea/agora/issues/1297#issuecomment-742203550

Then full nodes which want to validate really old blocks where those validators don't exist anymore would just use the `getPreimage` API to find those outdated validator's last preimages.

The problem right now is that we delete items from the `validator_set` as soon as a node gets un-enrolled. But we should probably keep that data forever so historic blocks can be validated. Of course, you would still need some information about which validators are **inactive**. Currently they're marked inactive by removing them from the validator set database, but maybe we should instead just add a status flag instead?

Reply:
> Currently they're marked inactive by removing them from the validator set database, but maybe we should just add a status flag instead?

The state could solve my problem. But we should think about the situation that the data of `validator_set` will be ever-increasing, which might not be so big though.